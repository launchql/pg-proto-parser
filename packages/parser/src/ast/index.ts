export * from './enums';
export * from './types';
export * from './generators'