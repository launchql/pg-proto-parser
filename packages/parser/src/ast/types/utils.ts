import * as t from '@babel/types';

export const getTSType = (type: string = 'any') => {
  switch (type) {
    case 'string':
      return t.tsStringKeyword();
    case 'double':
    case 'float':
    case 'int32':
    case 'uint32':
    case 'sint32':
    case 'fixed32':
    case 'sfixed32':
      return t.tsNumberKeyword();
    case 'int64':
    case 'uint64':
    case 'sint64':
    case 'fixed64':
    case 'sfixed64':
      return t.tsBigIntKeyword()
    case 'bytes':
      return t.tsTypeReference(t.identifier('Uint8Array'));
    case 'bool':
      return t.tsBooleanKeyword();
    default:
      // For custom types, reference them directly
      return t.tsTypeReference(t.identifier(type));

  };
};

export const isPrimitiveType = (type: string) => {
  switch (type) {
    case 'string':
    case 'double':
    case 'float':
    case 'int32':
    case 'uint32':
    case 'sint32':
    case 'fixed32':
    case 'sfixed32':
    case 'int64':
    case 'uint64':
    case 'sint64':
    case 'fixed64':
    case 'sfixed64':
    case 'bytes':
    case 'bool':
      return true;
    default:
      return false;
  };
};

export const resolveTypeName = (type: string) => {
  if (isPrimitiveType(type)) {
    return getTSType(type);
  }
  return t.tsTypeReference(t.identifier(type));
}
