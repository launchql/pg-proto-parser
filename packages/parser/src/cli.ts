#!/usr/bin/env node

import { PgProtoParser } from './parser';
import { program } from 'commander';
import { join } from 'path';

program
  .name('pg-proto-parser')
  .description('PostgreSQL protobuf parser and code generator')
  .version('1.0.0');

program
  .command('generate')
  .description('Generate TypeScript types and runtime schema from protobuf files')
  .argument('<proto-file>', 'Path to the protobuf file')
  .option('-o, --out-dir <dir>', 'Output directory', './out')
  .option('--types', 'Generate TypeScript types')
  .option('--enums', 'Generate TypeScript enums')
  .option('--runtime-schema', 'Generate runtime schema')
  .option('--runtime-schema-format <format>', 'Runtime schema format (json|typescript)', 'json')
  .action((protoFile, options) => {
    const parserOptions = {
      outDir: options.outDir,
      types: {
        enabled: options.types || false,
        filename: 'types.ts'
      },
      enums: {
        enabled: options.enums || false,
        filename: 'enums.ts'
      },
      runtimeSchema: {
        enabled: options.runtimeSchema || false,
        filename: 'runtime-schema',
        format: options.runtimeSchemaFormat
      }
    };

    try {
      const parser = new PgProtoParser(protoFile, parserOptions);
      parser.write();
      console.log(`Generated files in ${options.outDir}`);
    } catch (error) {
      console.error('Error generating files:', error);
      process.exit(1);
    }
  });

program.parse();
