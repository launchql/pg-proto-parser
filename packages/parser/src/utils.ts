import { Enum, Field } from '@launchql/protobufjs';

export const getUndefinedKey = (enumName) => {
  // Split the name into parts where a lowercase letter is followed by an uppercase letter
  const parts = enumName.split(/(?<=[a-z])(?=[A-Z])/);

  const processedParts = parts.map(part => {
    // For parts that are all uppercase and longer than 1 character, only the first character should remain uppercase
    if (part === part.toUpperCase() && part.length > 1) {
      return part.charAt(0) + part.slice(1).toLowerCase();
    }
    return part;
  });

  const upperSnakeCase = processedParts.join('_').toUpperCase();

  return `${upperSnakeCase}_UNDEFINED`;
}


export const hasUndefinedInitialValue = (enumData: Enum) => {
  const entries = Object.entries(enumData.values);
  if (entries.length === 0) return false;

  const undefinedKey = getUndefinedKey(enumData.name);
  const firstEntry = entries[0];
  return firstEntry[0] === undefinedKey && firstEntry[1] === 0;
}

export const getFieldName = (field: Field, fallbackName: string) => {
 return field.options?.json_name ? field.options.json_name : fallbackName;
}