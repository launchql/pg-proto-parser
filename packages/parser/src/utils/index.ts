export * from './proto';
export * from './files';
export * from './imports';
export * from './generate';
export * from './deps';
export * from './types';
export * from './meta';