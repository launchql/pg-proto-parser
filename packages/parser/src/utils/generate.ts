import generate from '@babel/generator';
import * as t from '@babel/types';

export const convertAstToCode = (body: any[]) => {
  const ast = t.file(t.program(body));
  const { code } = generate(ast);
  return code;
};