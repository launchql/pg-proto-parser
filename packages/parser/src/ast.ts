import { Service, Type, Field, Enum, Root, Namespace, ReflectionObject } from '@launchql/protobufjs';
import * as t from '@babel/types';
import generate from '@babel/generator';
import { getFieldName } from './utils';

export const getTSType = (type: string) => {
  switch (type) {
    case 'string':
      return t.tsStringKeyword();
    case 'double':
    case 'float':
    case 'int32':
    case 'uint32':
    case 'sint32':
    case 'fixed32':
    case 'sfixed32':
      return t.tsNumberKeyword();
    case 'int64':
    case 'uint64':
    case 'sint64':
    case 'fixed64':
    case 'sfixed64':
      return t.tsBigIntKeyword()
    case 'bytes':
      return t.tsTypeReference(t.identifier('Uint8Array'));
    case 'bool':
      return t.tsBooleanKeyword();
    default:
      throw new Error('getTSType() type not found');
  };
};

export const isPrimitiveType = (type: string) => {
  switch (type) {
    case 'string':
    case 'double':
    case 'float':
    case 'int32':
    case 'uint32':
    case 'sint32':
    case 'fixed32':
    case 'sfixed32':
    case 'int64':
    case 'uint64':
    case 'sint64':
    case 'fixed64':
    case 'sfixed64':
    case 'bytes':
    case 'bool':
      return true;
    default:
      return false;
  };
};

export const resolveTypeName = (type: string) => {
  if (isPrimitiveType(type)) {
    return getTSType(type);
  }
  return t.tsTypeReference(t.identifier(type));
}

export const transformEnumToAST = (enumData) => {
  const members = Object.entries(enumData.values).map(([key, value]) =>
    t.tsEnumMember(t.identifier(key), t.numericLiteral(value as number))
  );

  const enumDeclaration = t.tsEnumDeclaration(t.identifier(enumData.name), members);
  return t.exportNamedDeclaration(enumDeclaration);
};

export const transformTypeToAST = (type: Type) => {
  const typeName = type.name;
  // @ts-ignore
  const fields: Field[] = type.fields;
  const properties =
    Object.entries(fields)
      .map(([fieldName, fieldData]) => {
        const type = resolveTypeName(fieldData.type);
        const fieldType = fieldData.rule === 'repeated' ?
          t.tsArrayType(type) :
          type;
        return t.tsPropertySignature(t.identifier(getFieldName(fieldData, fieldName)), t.tsTypeAnnotation(fieldType));
      });

  const interfaceDecl = t.tsInterfaceDeclaration(
    t.identifier(typeName),
    null,
    [],
    t.tsInterfaceBody(properties)
  );

  // Wrap the interface declaration in an export statement
  return t.exportNamedDeclaration(interfaceDecl, []);
}

export const generateTSInterfaces = (types: Type[]) => {
  const ast = t.file(t.program(types.map(type => transformTypeToAST(type))));
  const { code } = generate(ast);
  return code;
}

export const generateTSEnums = (enums: Enum[]) => {
  const ast = t.file(t.program(enums.map(enm => transformEnumToAST(enm))));
  const { code } = generate(ast);
  return code;
}

export const generateTSEnumFunction = (enums: Enum[]) => {
  const ast = t.file(t.program(buildEnumValueFunctionAST(enums)));
  const { code } = generate(ast);
  return code;
}



export const buildEnumValueFunctionAST = (enumData: Enum[]) => {
  // Create the union type for EnumType
  const enumTypeIdentifier = t.identifier('EnumType');
  const enumTypeUnion = t.tsUnionType(
    enumData.map(enumDef => t.tsLiteralType(t.stringLiteral(enumDef.name)))
  );
  const enumTypeAlias = t.tsTypeAliasDeclaration(enumTypeIdentifier, null, enumTypeUnion);

  // Create the function parameter and its type
  const enumTypeParam = t.identifier('enumType');
  const enumTypeParamAnnotation = t.tsTypeAnnotation(t.tsTypeReference(enumTypeIdentifier));
  enumTypeParam.typeAnnotation = enumTypeParamAnnotation;

  const keyParam = t.identifier('key');
  const keyParamAnnotation = t.tsTypeAnnotation(
    t.tsUnionType([t.tsStringKeyword(), t.tsNumberKeyword()])
  );
  keyParam.typeAnnotation = keyParamAnnotation;

  // Create the switch cases for the outer switch statement
  const outerCases = enumData.map(enumDef => {
    const innerCases = Object.entries(enumDef.values).map(([key, value]) => {
      return t.switchCase(t.stringLiteral(key), [t.returnStatement(t.numericLiteral(value))]);
    });

    // Add the reverse mapping cases
    innerCases.push(...Object.entries(enumDef.values).map(([key, value]) => {
      return t.switchCase(t.numericLiteral(value), [t.returnStatement(t.stringLiteral(key))]);
    }));

    // Add the default case for the inner switch
    innerCases.push(t.switchCase(null, [
      t.throwStatement(t.newExpression(t.identifier('Error'), [t.stringLiteral(`Key not recognized in enum ${enumDef.name}`)]))
    ]));

    return t.switchCase(t.stringLiteral(enumDef.name), [
      t.blockStatement([t.switchStatement(keyParam, innerCases)])
    ]);
  });

  // Add the default case for the outer switch
  outerCases.push(t.switchCase(null,
    [
      t.throwStatement(
        t.newExpression(t.identifier('Error'), [t.stringLiteral('Enum type not recognized')])
      )
    ]
  ));

  // Create the outer switch statement
  const switchStatement = t.switchStatement(enumTypeParam, outerCases);

  // Create the function body
  const functionBody = t.blockStatement([switchStatement]);

  // Create the arrow function expression
  const arrowFunctionExpression = t.arrowFunctionExpression(
    [enumTypeParam, keyParam],
    functionBody
  );

  // Create the variable declarator for the function
  const variableDeclarator = t.variableDeclarator(
    t.identifier('getEnumValue'),
    arrowFunctionExpression
  );

  // Create the variable declaration (const getEnumValue = ...)
  const variableDeclaration = t.variableDeclaration('const', [variableDeclarator]);

  // Export the arrow function
  const exportedFunction = t.exportNamedDeclaration(variableDeclaration, []);

  // Export the type
  const exportedEnumTypeAlias = t.exportNamedDeclaration(enumTypeAlias, []);

  // Return the entire AST
  return [exportedEnumTypeAlias, exportedFunction];
}
