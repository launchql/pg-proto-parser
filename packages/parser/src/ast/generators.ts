import { Type, Enum } from '@launchql/protobufjs';
import * as t from '@babel/types';
import generate from '@babel/generator';
import { PgProtoParserOptions } from '../options';
import { createAstHelperMethodsAST, createUnionTypeAST, transformTypeToAST } from './types';
import { buildEnumValueFunctionAST, transformEnumToAST, transformEnumToTypeUnionAST } from './enums';

export const generateImportSpecifiersAST = (types: Type[], options: PgProtoParserOptions) => {
  const importSpecifiers = types.map(type =>
    t.importSpecifier(t.identifier(type.name), t.identifier(type.name))
  );

  const importDeclaration = t.importDeclaration(importSpecifiers, t.stringLiteral(options.astHelperTypeSource));
  return importDeclaration;
}

export const generateTSInterfaces = (
  types: Type[],
  options: PgProtoParserOptions,
  useNestedTypes: boolean
) => {
  const node = createUnionTypeAST(types.filter(type => type.name !== 'Node'));
  const typeDefns = types.reduce((m, type) => {
    if (type.name === 'Node') return m;
    return [...m, transformTypeToAST(type, options, useNestedTypes)]
  }, []);

  const ast = t.file(t.program(
    [
      node,
      ...typeDefns
    ]
  ));
  const { code } = generate(ast);
  return code;
}

export const generateTSEnums = (enums: Enum[]) => {
  const ast = t.file(t.program(enums.map(enm => transformEnumToAST(enm))));
  const { code } = generate(ast);
  return code;
}

export const generateTSEnumsTypeUnionAST = (enums: Enum[]) => {
  const ast = t.file(t.program(enums.map(enm => transformEnumToTypeUnionAST(enm))));
  const { code } = generate(ast);
  return code;
}

export const generateTSEnumFunction = (enums: Enum[]) => {
  const ast = t.file(t.program(buildEnumValueFunctionAST(enums)));
  const { code } = generate(ast);
  return code;
}
export const generateTSASTHelperMethods = (types: Type[]) => {
  const ast = t.file(t.program([
    // generateMacroForTypes(),
    createAstHelperMethodsAST(types)
  ]));
  const { code } = generate(ast);
  return code;
}

export const generateTSASTHelpersImports = (types: Type[], options: PgProtoParserOptions) => {
  const ast = t.file(t.program([generateImportSpecifiersAST(types, options)]));
  const { code } = generate(ast);
  return code;
};