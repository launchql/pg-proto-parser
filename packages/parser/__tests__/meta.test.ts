import ast from '../test-utils/utils/asts';
import { generateTsAstCodeFromPgAst } from '../src/utils'
import generate from '@babel/generator';

it('AST to AST to create AST â€”Â meta ðŸ¤¯', () => {
    const selectStmt = ast.selectStmt({
        targetList: [
          ast.resTarget({
            val: ast.columnRef({
              fields: [ast.aStar()]
            })
          })
        ],
        fromClause: [
          ast.rangeVar({
            relname: 'some_amazing_table',
            inh: true,
            relpersistence: 'p'
          })
        ],
        limitOption: 'LIMIT_OPTION_DEFAULT',
        op: 'SETOP_NONE'
      });

      expect(selectStmt).toMatchSnapshot();
      
      const astForAst = generateTsAstCodeFromPgAst(selectStmt);
      expect(generate(astForAst).code).toMatchSnapshot();
      expect(generate(astForAst).code).toEqual(
`ast.selectStmt({
  targetList: [ast.resTarget({
    val: ast.columnRef({
      fields: [ast.aStar({})]
    })
  })],
  fromClause: [ast.rangeVar({
    relname: "some_amazing_table",
    inh: true,
    relpersistence: "p"
  })],
  limitOption: "LIMIT_OPTION_DEFAULT",
  op: "SETOP_NONE"
})`);

})

it('works', () => {
    const parsedResult = [
        {
          "RawStmt": {
            "stmt": {
              "SelectStmt": {
                "targetList": [
                  {
                    "ResTarget": {
                      "val": {
                        "ColumnRef": {
                          "fields": [
                            {
                              "A_Star": {}
                            }
                          ],
                          "location": 7
                        }
                      },
                      "location": 7
                    }
                  }
                ],
                "fromClause": [
                  {
                    "RangeVar": {
                      "relname": "my_table",
                      "inh": true,
                      "relpersistence": "p",
                      "location": 14
                    }
                  }
                ],
                "whereClause": {
                  "A_Expr": {
                    "kind": "AEXPR_OP",
                    "name": [
                      {
                        "String": {
                          "str": "="
                        }
                      }
                    ],
                    "lexpr": {
                      "ColumnRef": {
                        "fields": [
                          {
                            "String": {
                              "str": "id"
                            }
                          }
                        ],
                        "location": 29
                      }
                    },
                    "rexpr": {
                      "A_Const": {
                        "val": {
                          "Integer": {
                            "ival": 1
                          }
                        },
                        "location": 34
                      }
                    },
                    "location": 32
                  }
                },
                "limitOption": "LIMIT_OPTION_DEFAULT",
                "op": "SETOP_NONE"
              }
            },
            "stmt_location": 0
          }
        }
      ];

      const astForAst = generateTsAstCodeFromPgAst(parsedResult[0].RawStmt.stmt);
      expect(generate(astForAst).code).toMatchSnapshot();

})