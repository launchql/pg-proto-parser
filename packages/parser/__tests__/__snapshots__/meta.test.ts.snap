// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AST to AST to create AST â€”Â meta ðŸ¤¯ 1`] = `
{
  "SelectStmt": {
    "fromClause": [
      {
        "RangeVar": {
          "inh": true,
          "relname": "some_amazing_table",
          "relpersistence": "p",
        },
      },
    ],
    "limitOption": "LIMIT_OPTION_DEFAULT",
    "op": "SETOP_NONE",
    "targetList": [
      {
        "ResTarget": {
          "val": {
            "ColumnRef": {
              "fields": [
                {
                  "A_Star": {},
                },
              ],
            },
          },
        },
      },
    ],
  },
}
`;

exports[`AST to AST to create AST â€”Â meta ðŸ¤¯ 2`] = `
"ast.selectStmt({
  targetList: [ast.resTarget({
    val: ast.columnRef({
      fields: [ast.aStar({})]
    })
  })],
  fromClause: [ast.rangeVar({
    relname: "some_amazing_table",
    inh: true,
    relpersistence: "p"
  })],
  limitOption: "LIMIT_OPTION_DEFAULT",
  op: "SETOP_NONE"
})"
`;

exports[`works 1`] = `
"ast.selectStmt({
  targetList: [ast.resTarget({
    val: ast.columnRef({
      fields: [ast.aStar({})],
      location: 7
    }),
    location: 7
  })],
  fromClause: [ast.rangeVar({
    relname: "my_table",
    inh: true,
    relpersistence: "p",
    location: 14
  })],
  whereClause: ast.aExpr({
    kind: "AEXPR_OP",
    name: [ast.string({
      str: "="
    })],
    lexpr: ast.columnRef({
      fields: [ast.string({
        str: "id"
      })],
      location: 29
    }),
    rexpr: ast.aConst({
      val: ast.integer({
        ival: 1
      }),
      location: 34
    }),
    location: 32
  }),
  limitOption: "LIMIT_OPTION_DEFAULT",
  op: "SETOP_NONE"
})"
`;
